examples/amm/src/contract.rs

```diff
@@ -0,0 +1,342 @@
+                    let new_ratio_bigint =
+                        (token0_amount_bigint * one_amount_bigint) / token1_amount_bigint;
+
+                    if balance1 > Amount::ZERO && old_ratio_bigint != new_ratio_bigint {
+                        return Err(AmmError::BalanceRatioAlteredError);
+                    }
+                }
+
+                self.receive_from_account(&owner, 0, token0_amount).await?;
+                self.receive_from_account(&owner, 1, token1_amount).await?;
+
+                Ok(())
+            }
+            // When removing liquidity, you'll specify one of the tokens you want to
+            // remove and the amount, and we'll calculate the amount for the other token that
+            // we'll remove based on the current ratio, and remove them.
+            OperationType::RemoveLiquidity {
+                owner,
+                input_token_idx,
+                input_amount,
+                other_token_idx,
+            } => {
+                if input_token_idx > 1 || other_token_idx > 1 {
+                    return Err(AmmError::InvalidTokenIdx);
+                }
+
+                if input_token_idx == other_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let balance0 = self.get_pool_balance(0).await?;
+                let balance1 = self.get_pool_balance(1).await?;
+
+                if (input_token_idx == 0 && input_amount > balance0)
+                    || (input_token_idx == 1 && input_amount > balance1)
+                {
+                    return Err(AmmError::InsufficientLiquidityError);
+                }
+
+                let one_amount_bigint =
+                    BigUint::from_u128(<Amount as Into<u128>>::into(Amount::ONE))
+                        .expect("Couldn't generate Amount::ONE in bigint");
+
+                let input_amount_bigint =
+                    BigUint::from_u128(<Amount as Into<u128>>::into(input_amount))
+                        .expect("Couldn't generate input_amount in bigint");
+
+                let balance0_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(balance0))
+                    .expect("Couldn't generate balance0 in bigint");
+                let balance1_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(balance1))
+                    .expect("Couldn't generate balance1 in bigint");
+
+                // To always maintain the precision, whenever we multiply we divide by Amount::ONE
+                // Whenever we divide, we always multiply by Amount::ONE before dividing
+                // The final division by Amount::ONE is to round down to the closest integer
+                let other_amount = if input_token_idx == 0 {
+                    Amount::from_tokens(
+                        (((((input_amount_bigint * balance1_bigint) / one_amount_bigint.clone())
+                            * one_amount_bigint.clone())
```

Comment:
Why divide and then multiply by `ONE`?

Reply:
Was for the fixed precision stuff that we figured out is unnecessary, but will be fixed in the next PR update!